packages/jaspr/lib/src/browser/slotted_dom_nodes_component.dart

```diff
@@ -0,0 +1,236 @@
+
+  // SlottedChildView._withNodes({required List<web.Node> this.nodes, required this.slots});
+
+  final List<web.Node>? nodes;
+  final List<ChildSlot> slots;
+
+  @override
+  Element createElement() => SlottedDomNodesElement(this);
+}
+
+class SlottedDomNodesElement extends MultiChildRenderObjectElement {
+  SlottedDomNodesElement(SlottedChildView super.component);
+
+  @override
+  SlottedChildView get component => super.component as SlottedChildView;
+
+  @override
+  void update(SlottedChildView newComponent) {
+    assert(
+      newComponent.nodes == component.nodes,
+      'SlottedDomNodesComponent cannot be updated with different nodes.',
+    );
+    super.update(newComponent);
+  }
+
+  @override
+  List<Component> buildChildren() {
+    return component.slots;
+  }
+
+  @override
+  RenderObject createRenderObject() {
+    final parent = parentRenderObjectElement!.renderObject;
+    return SlottedDomRenderObject.fromNodes(component.nodes, parent as DomRenderObject);
+  }
+
+  @override
+  void updateRenderObject(RenderObject renderObject) {}
+
+  @override
+  void unmount() {
+    super.unmount();
+    _clearEventListeners(this);
+  }
+
+  static _clearEventListeners(Element e) {
+    if (e case RenderObjectElement(renderObject: DomRenderElement r)) {
+      r.events?.forEach((type, binding) {
+        binding.clear();
+      });
+      r.events = null;
+    }
+
+    e.visitChildren(_clearEventListeners);
+  }
+}
+
+class SlottedDomRenderObject extends DomRenderFragment {
+  SlottedDomRenderObject(DomRenderObject? parent) : super(parent, []);
+
+  factory SlottedDomRenderObject.fromNodes(List<web.Node>? nodes, DomRenderObject? parent) {
+    nodes ??= (parent is HydratableDomRenderObject) ? [...parent.toHydrate] : <web.Node>[];
+
+    final object = SlottedDomRenderObject(parent);
+    if (nodes.isEmpty) {
+      object.firstChildNode = null;
+      object.lastChildNode = null;
+      object.isAttached = true;
+    } else {
+      final firstNode = nodes.first;
+      final lastNode = nodes.last;
+      assert(firstNode.parentNode == lastNode.parentNode, 'All nodes must share the same parent.');
+      object.firstChildNode = firstNode;
+      object.lastChildNode = lastNode;
+      if (parent != null && parent.node.contains(firstNode)) {
+        if (parent is HydratableDomRenderObject) {
+          var startIndex = parent.toHydrate.indexOf(firstNode);
+          var endIndex = parent.toHydrate.indexOf(lastNode);
+          if (startIndex != -1 && endIndex != -1 && startIndex <= endIndex) {
+            parent.toHydrate.removeRange(startIndex, endIndex + 1);
+          }
+        }
+        object.isAttached = true;
+      } else {
+        for (final node in nodes) {
+          object.node.appendChild(node);
+        }
+      }
+    }
+    return object;
+  }
+
+  @override
+  late final web.Node? firstChildNode;
+
+  @override
+  late final web.Node? lastChildNode;
```

Comment:
Seems these could be set in the constructor to avoid being `late`.

Reply:
They are, but require too much logic for initializers.